vllm/attention/backends/pallas.py

```diff
@@ -296,16 +297,26 @@ def write_to_kv_cache(
     key_cache: torch.Tensor,
     value_cache: torch.Tensor,
     slot_mapping: torch.Tensor,
+    is_prefill: bool,
 ) -> None:
     torch.ops.xla.dynamo_set_buffer_donor_(key_cache, True)
     torch.ops.xla.dynamo_set_buffer_donor_(value_cache, True)
 
-    key = key.flatten(0, 2)
-    value = value.flatten(0, 2)
-    key_cache = key_cache.flatten(0, 2)
-    value_cache = value_cache.flatten(0, 2)
-    key_cache.index_copy_(0, slot_mapping, key)
-    value_cache.index_copy_(0, slot_mapping, value)
+    if is_prefill:
+        bs, sq, hn, hs = key.shape
+        key = key.view(sq // MIN_PREFILL_SEQ_LEN, MIN_PREFILL_SEQ_LEN, hn, hs)
+        value = value.view(sq // MIN_PREFILL_SEQ_LEN, MIN_PREFILL_SEQ_LEN, hn, hs)
+        slot_mapping = slot_mapping.flatten()
+        key_cache.index_copy_(0, slot_mapping, key)
+        value_cache.index_copy_(0, slot_mapping, value)
+    else:
+        key = key.flatten(0, 1)
+        value = value.flatten(0, 1)
+        key_cache = key_cache.flatten(0, 1)
+        value_cache = value_cache.flatten(0, 1)
+        slot_mapping = slot_mapping.flatten()
```

Comment:
I wonder why you need to flatten the slot_mapping here

Reply:
It's a 2D tensor before that